src/li_chao.rs

```diff
@@ -0,0 +1,99 @@
+/// A structure for answering minimum queries on a set of linear functions. Supports two
+/// operations: inserting a linear function and querying for minimum at a given point. Unlike the
+/// simplest convex hull trick implementation, the queries can be done in any order, and we can do
+/// all the calculations using integers.
+
+/// This data structure builds a static segment tree over the interval of x-coordinates
+/// [left,right), so requires O(right-left) memory. Just like normal segment trees, this could be
+/// modified to a dynamic tree when the range is huge (it can also be made persistent!).
+/// It can also probably be done as an AlCash style iterative segment tree in the static case.
+
+pub struct LiChaoTree {
+    left: i64,
+    right: i64,
+    lines: Vec<(i64, i64)>,
+}
+
+impl LiChaoTree {
+    /// Creates a new tree, built to handle queries on the interval [left, right).
+    pub fn new(left: i64, right: i64) -> Self {
+        Self {
+            left,
+            right,
+            lines: vec![(0, i64::MAX); 4 * (right - left + 1) as usize],
+        }
+    }
+
+    /// Every node in the tree has the property that the line that minimizes its midpoint is found
+    /// either in the node or one of its ancestors.  When we visit a node, we compute the winner at
+    /// the midpoint of the node. The winner is stored in the node. The loser can still possibly
+    /// beat the winner on some segment, either to the left or to the right of the current
+    /// midpoint, so we propagate it to that segment. This sequence ensures that the invariant is
+    /// kept.
+    fn add_line_impl(&mut self, mut m: i64, mut b: i64, ix: usize, l: i64, r: i64) {
```

Comment:
It would involve some funky transformations, but I believe it's possible to use `dynamic_arq::ArqView` instead of `ix, l, r` if you feel so inclined :P

Reply:
Actually, better yet, is it possible to just use `r-l+1` nodes instead of four times that? There's no need for the subtree to again include the middle node, right? So instead, you'd start your search in the middle node, then go either left or right. Anything wrong with this idea?

Reply:
I am not sure what you mean about including the middle node. Is there a way to get rid of some of the nodes? I think they are all needed.

The standard bound for segment tree nodes is 4N; if N is a power of 2 it's 2N but I think we need around 4N if N is slightly higher than a power of 2. That's one of the reasons the AlCash style is more efficient. That one has a bound of 2N regardless of N. But I don't know if it's easy to convert Li Chao trees to that style - they are inherently top-down, moreso than normal segment trees.

Converting to use the ArqView interface would be an interesting puzzle... I wonder if it can be done while keeping time/space complexity.

Reply:
Space optimization added in latest update.

Reply:
Yay!